File: packages/python/anip-server/src/anip_server/audit.py
"""Audit log manager for ANIP services."""
from __future__ import annotations

import inspect
from collections.abc import Awaitable
from datetime import datetime, timezone
from typing import Any, Callable

from .storage import StorageBackend


class AuditLog:
    """Audit log backed by a StorageBackend.

    Sequence numbers, hash chaining, and Merkle accumulation are now
    handled by the storage layer (``append_audit_entry``).
    """

    def __init__(
        self,
        storage: StorageBackend,
        signer: Callable[[dict[str, Any]], str | Awaitable[str]] | None = None,
    ) -> None:
        self._storage = storage
        self._signer = signer

    async def log_entry(self, entry_data: dict[str, Any]) -> dict[str, Any]:
        """Log an audit entry via the storage backend.

        entry_data should contain: capability, token_id, root_principal, success,
        and optionally: issuer, subject, parameters, result_summary, failure_type,
        cost_actual, delegation_chain.

        Returns the complete entry dict (with sequence_number, timestamp, previous_hash, signature).
        """
        now = datetime.now(timezone.utc).isoformat()
        entry_for_storage = {
            "timestamp": now,
            "capability": entry_data["capability"],
            "token_id": entry_data.get("token_id"),
            "issuer": entry_data.get("issuer"),
            "subject": entry_data.get("subject"),
            "root_principal": entry_data.get("root_principal"),
            "parameters": entry_data.get("parameters"),
            "success": entry_data["success"],
            "result_summary": entry_data.get("result_summary"),
            "failure_type": entry_data.get("failure_type"),
            "cost_actual": entry_data.get("cost_actual"),
            "delegation_chain": entry_data.get("delegation_chain"),
            "invocation_id": entry_data.get("invocation_id"),
            "client_reference_id": entry_data.get("client_reference_id"),
            "task_id": entry_data.get("task_id"),
            "parent_invocation_id": entry_data.get("parent_invocation_id"),
            "stream_summary": entry_data.get("stream_summary"),
            "event_class": entry_data.get("event_class"),
            "retention_tier": entry_data.get("retention_tier"),
            "expires_at": entry_data.get("expires_at"),
            "storage_redacted": entry_data.get("storage_redacted", False),
            "entry_type": entry_data.get("entry_type"),
            "grouping_key": entry_data.get("grouping_key"),
            "aggregation_window": entry_data.get("aggregation_window"),
            "aggregation_count": entry_data.get("aggregation_count"),
            "first_seen": entry_data.get("first_seen"),
            "last_seen": entry_data.get("last_seen"),
            "representative_detail": entry_data.get("representative_detail"),
        }

        entry = await self._storage.append_audit_entry(entry_for_storage)

        if self._signer:
            sig = self._signer(entry)
            if inspect.isawaitable(sig):
                sig = await sig
            entry["signature"] = sig
            await self._storage.update_audit_signature(entry["sequence_number"], sig)
        else:
            entry["signature"] = None

        return entry

    async def query(self, **filters: Any) -> list[dict[str, Any]]:
        """Query audit entries with optional filters."""
        return await self._storage.query_audit_entries(**filters)

File: packages/typescript/server/src/audit.ts
/**
 * Audit log manager for ANIP services.
 *
 * Sequence numbers, hash chaining, and Merkle accumulation are now
 * handled by the storage layer ({@link StorageBackend.appendAuditEntry}).
 */

import type { StorageBackend } from "./storage.js";

export class AuditLog {
  private _storage: StorageBackend;
  private _signer: ((entry: Record<string, unknown>) => string | Promise<string>) | null;

  constructor(
    storage: StorageBackend,
    signer?: (entry: Record<string, unknown>) => string | Promise<string>,
  ) {
    this._storage = storage;
    this._signer = signer ?? null;
  }

  /**
   * Log an audit entry via the storage backend.
   *
   * `entryData` should contain: `capability`, `token_id`, `root_principal`,
   * `success`, and optionally: `issuer`, `subject`, `parameters`,
   * `result_summary`, `failure_type`, `cost_actual`, `delegation_chain`.
   *
   * Returns the complete entry dict (with `sequence_number`, `timestamp`,
   * `previous_hash`, `signature`).
   */
  async logEntry(entryData: Record<string, unknown>): Promise<Record<string, unknown>> {
    const now = new Date().toISOString();

    const entryForStorage: Record<string, unknown> = {
      timestamp: now,
      capability: entryData.capability,
      token_id: entryData.token_id ?? null,
      issuer: entryData.issuer ?? null,
      subject: entryData.subject ?? null,
      root_principal: entryData.root_principal ?? null,
      parameters: entryData.parameters ?? null,
      success: entryData.success,
      result_summary: entryData.result_summary ?? null,
      failure_type: entryData.failure_type ?? null,
      cost_actual: entryData.cost_actual ?? null,
      delegation_chain: entryData.delegation_chain ?? null,
      invocation_id: entryData.invocation_id ?? null,
      client_reference_id: entryData.client_reference_id ?? null,
      task_id: entryData.task_id ?? null,
      parent_invocation_id: entryData.parent_invocation_id ?? null,
      upstream_service: entryData.upstream_service ?? null,
      stream_summary: entryData.streamSummary ?? entryData.stream_summary ?? null,
      event_class: entryData.event_class ?? null,
      retention_tier: entryData.retention_tier ?? null,
      expires_at: entryData.expires_at ?? null,
      storage_redacted: entryData.storage_redacted ?? false,
      entry_type: entryData.entry_type ?? null,
      grouping_key: entryData.grouping_key ?? null,
      aggregation_window: entryData.aggregation_window ?? null,
      aggregation_count: entryData.aggregation_count ?? null,
      first_seen: entryData.first_seen ?? null,
      last_seen: entryData.last_seen ?? null,
      representative_detail: entryData.representative_detail ?? null,
      // v0.23 — approval flow linkage. See SPEC.md §4.7–§4.9.
      approval_request_id: entryData.approval_request_id ?? null,
      approval_grant_id: entryData.approval_grant_id ?? null,
    };

    const entry = await this._storage.appendAuditEntry(entryForStorage);

    entry.signature = this._signer ? await this._signer(entry) : null;
    if (entry.signature) {
      await this._storage.updateAuditSignature(entry.sequence_number as number, entry.signature as string);
    }

    return entry;
  }

  /** Query audit entries with optional filters. */
  async query(opts?: {
    capability?: string;
    rootPrincipal?: string;
    since?: string;
    invocationId?: string;
    clientReferenceId?: string;
    taskId?: string;
    parentInvocationId?: string;
    eventClass?: string;
    limit?: number;
  }): Promise<Record<string, unknown>[]> {
    return await this._storage.queryAuditEntries(opts);
  }
}
